### ISSUE TITLE
Allow File System Provider workspaces without loosing untitled documents

### ISSUE THREAD

**Issue by grork** [Microsoft/vscode#50013]
Scenario:
1. Have a File System Provider that doesn't use any local files & handles everything in memory
2. "Open" a workspace using that file system provider
3. Create untitled documents
4. Programmatically close the workspace

Right now, at 4, the customer will be prompted to save the untitled documents -- which they can cancel and, thus cause the closing of the space to be "Aborted".

In our specific scenario, there are situations where the source of our data for the File SYstem provider "goes away", and it's much cleaner for us to have them close the workspace than allow them to continue to operating like things are "working".

Looking into "hotExit" behaviour, it _almost_ gives us the behaviour we want (no loss of untitled documents), but will recover unsaved documents (not untitled), which we handle in a different way.

Another option for us is a way to close the workspace, without triggering a reload.

**Comment by bpasero** [comment]
@grork I do not see how file system providers and untitled documents are related, you should never get prompted when closing with untitled documents independent from whatever extension is installed. Could it be that you are using LiveShare where hot exit is explicitly disabled?

**Comment by bpasero** [comment]
@grork I am having a hard time following you, imho hot-exit has one task and that is to preserve whatever the user typed and then restore it at the next startup. It does not make any difference if the file is untitled or not as long as the file is dirty and thus contains things that the user has typed. Again, can you please explain why you think file system providers play any role in how hot exit should behave?

**Comment by grork** [comment]
Sorry for not being clear, I apologise. Let me try again.

In non-file system provider scenarios, there is a (almost) guarantee that the disk doesn't randomly disappear for reasons that the user didn't initiate.

With a file system provider, the "disk" that those files are stored on, is transient. It might need to go away for many reasons (network, or permissions loss on the remote side for example).

This presents a challenge for the file system provider implementer -- when the backing store for it goes away, the _cleanest_ solution seems to be to close the workspace. This makes it easy on the provider as all those corner cases with it still being in use go away (e.g. having to remove file actions in the explorer tree, and emptying the tree by raising file change notifications)

This leads to the question: When programmatically closing the workspace, we can avoid the cancel path by saving the edited documents. This results in the workspace closing, and we're good. _However_ that doesn't apply to "Untitled" documents. This means when the workspace is closing, they prompted for untitled documents, which they can cancel which returns us to the situation of having to "Block" operations on the file system provider.

We don't want to lose untitled document data -- e.g. doing a revert/undo to empty state/etc seems _bad_ from a data loss perspective.

Which leads us to HotExit. An excellent solution to this case. Except when it handles _all_ files in the workspace, and results in VS Code attempting to open the workspace again (the reason for it to exist has now passed, so thats bad), and when it does that it auto-restores _unsaved_ documents (not just untitled). We don't want need that either.

Options I see:
a) Mark documents individually to not participate in hotExit? (I didn't see an API for this, maybe I missed it)
b) Mark the workspace it self as not participating in the hotExit (I don't see that either)
c) Option for us to save the untitled documents to the users real file system without them being prompted / cancelled. (Didn't see that either)
d) Option to close a workspace that will re-open the the untitled documents when window reloads.
d) An option on HotExit that says "I only care about untitled documents"

**Comment by bpasero** [comment]
> With a file system provider, the "disk" that those files are stored on, is transient. It might need to go away for many reasons (network, or permissions loss on the remote side for example).
> 
> This presents a challenge for the file system provider implementer -- when the backing store for it goes away, the cleanest solution seems to be to close the workspace. This makes it easy on the provider as all those corner cases with it still being in use go away (e.g. having to remove file actions in the explorer tree, and emptying the tree by raising file change notifications)

This seems to be the premise to the entire discussion and I question that. Why would you close the workspace if a remote file system goes away? The user could be in a setup where multiple of the workspace root folders are connected to multiple different remote file systems. If one goes away, you cannot just close the workspace because the other workspace roots might still work perfectly fine. 

Why would the user not simply be able to continue using the workspace, even if the file system goes away? I think what could happen is:
* e.g. the explorer shows a decoration on the workspace root that the connection was lost and so the tree is no longer functional for this workspace in the file explorer
* any unsaved document that was opened from this file system provider will continue to work as before, but you cannot save it
* any opened document that is not dirty will now throw an error as soon as you click on it. this is a behaviour that I am actually willing to revisit. Imho when a document was opened once and loaded into memory, we could still allow to open it again even if the underlying file store is gone

**Comment by bpasero** [comment]
> Another option for us is a way to close the workspace, without triggering a reload.

That is possible in a way. You could use the `updateWorkspaceFolders` API to remove the single root you have turning the window basically into an empty window. It would still be associated with the workspace file you generate in temp though, so not sure how much of an option that would be.

Alternatively you could:
* configure as workspace setting in your remote workspace: `"files.hotExit": "onExitAndWindowClose"` 
* close any dirty file (keeping the untitled files)

This allows to close the window without getting prompted and it will store dirty files as backup and restore them later.

**Comment by grork** [comment]
After much fighting, this ended up working. It's a OK middle ground -- it's a significant amount of work, and there are some corner cases that would be nice to support, but are a _lot_ of code on our side to support. But, I think this is good enough!
